crates/ruff/src/rules/isort/categorize.rs

```diff
@@ -239,34 +274,46 @@ impl KnownModules {
                 .rev()
             {
                 let submodule = &module_name[0..i];
-                if self.first_party.contains(submodule) {
-                    return Some((ImportType::FirstParty, Reason::KnownFirstParty));
-                }
-                if self.third_party.contains(submodule) {
-                    return Some((ImportType::ThirdParty, Reason::KnownThirdParty));
-                }
-                if self.local_folder.contains(submodule) {
-                    return Some((ImportType::LocalFolder, Reason::KnownLocalFolder));
-                }
-                if self.standard_library.contains(submodule) {
-                    return Some((ImportType::StandardLibrary, Reason::ExtraStandardLibrary));
+                if let Some(result) = self.categorize_submodule(submodule) {
+                    return Some(result);
                 }
             }
             None
         } else {
             // Happy path: no submodules, so we can check the module base and be done.
             let module_base = module_name.split('.').next().unwrap();
-            if self.first_party.contains(module_base) {
-                Some((ImportType::FirstParty, Reason::KnownFirstParty))
-            } else if self.third_party.contains(module_base) {
-                Some((ImportType::ThirdParty, Reason::KnownThirdParty))
-            } else if self.local_folder.contains(module_base) {
-                Some((ImportType::LocalFolder, Reason::KnownLocalFolder))
-            } else if self.standard_library.contains(module_base) {
-                Some((ImportType::StandardLibrary, Reason::ExtraStandardLibrary))
-            } else {
-                None
-            }
+            self.categorize_submodule(module_base)
+        }
+    }
+
+    fn categorize_submodule(&self, submodule: &str) -> Option<(ImportSection, Reason)> {
+        if let Some((section, _)) = self.sections.iter().find(|(_, v)| v.contains(submodule)) {
+            Some((
+                ImportSection::UserDefined(section.clone()),
+                Reason::UserDefinedSection,
+            ))
+        } else if self.first_party.contains(submodule) {
+            Some((
+                ImportSection::Known(ImportType::FirstParty),
+                Reason::KnownFirstParty,
+            ))
+        } else if self.third_party.contains(submodule) {
+            Some((
+                ImportSection::Known(ImportType::ThirdParty),
+                Reason::KnownThirdParty,
+            ))
+        } else if self.local_folder.contains(submodule) {
+            Some((
+                ImportSection::Known(ImportType::LocalFolder),
+                Reason::KnownLocalFolder,
+            ))
+        } else if self.standard_library.contains(submodule) {
+            Some((
+                ImportSection::Known(ImportType::StandardLibrary),
+                Reason::ExtraStandardLibrary,
+            ))
+        } else {
+            None
         }
```

Comment:
yes, just like there is currently nothing stopping a user from doing something like `extra-standard-library = ["path"]` and `known-first-party = ["path"]`. The first match found 'wins', and is the section that will be used. Configurations like that should probably result in an error or warning. Constructing a mapping sounds like a good idea, it would make this lookup  simpler and makes it easy to detect conflicts.